File: C++/56-Merge_Intervals.cpp
// Runtime: 76 ms, faster than 15.81% of C++ online submissions for Merge Intervals.
// Memory Usage: 28.3 MB, less than 5.17% of C++ online submissions for Merge Intervals.
class Solution {
public:
    vector<vector<int>> merge(vector<vector<int>>& intervals) {
        sort(intervals.begin(), intervals.end(), [](vector<int> it1, vector<int> it2) {
            return it1[0] < it2[0] || (it1[0] == it2[0] && it1[1] < it2[1]);
        });
        vector<vector<int>> res;
        for (vector<int> line : intervals)
            if (res.size() > 0 && line[0] <= res[res.size() - 1][1])
                res[res.size() - 1][1] = max(res[res.size() - 1][1], line[1]);
            else
                res.push_back(line);
        return res;
    }
};

File: Algorithms/Java/56-Merge_Intervals.java
// Runtime: 5 ms, faster than 94.78% of Java online submissions for Merge Intervals.
// Memory Usage: 41.7 MB, less than 67.51% of Java online submissions for Merge Intervals.
class Solution {
    public int[][] merge(int[][] intervals) {
        if (intervals.length <= 1) {
            return intervals;
        }
        Arrays.sort(intervals, (i1, i2) -> Integer.compare(i1[0], i2[0]));
        List<int[]> res = new ArrayList<>();
        int[] tmp = intervals[0];
        res.add(tmp);
        for (int[] interval : intervals) {
            if (interval[0] <= tmp[1]) {
                tmp[1] = Math.max(tmp[1], interval[1]);
            } else {
                tmp = interval;
                res.add(tmp);
            }
        }
        return res.toArray(new int[res.size()][]);
    }
}
